docs/0001-teleport-job-worker.md

```diff
@@ -0,0 +1,136 @@
+who created them. This will be done by adding `grpc.UnaryInterceptor` to the API server to check if the user has
+authorization to interact with the resource. For the purposes of this challenge, this type of authorization will be
+fine. However, proper OAuth would need to be implemented for use in production.
+
+### UX
+
+The CLI used in the service will be a simple application that implements the shared library and expose all the
+functionality of the [API](#api).
+
+For example, calling `Query()` in the API is equivalent to calling:
+
+```shell
+job-worker query some-user-token
+```
+
+An exhaustive list of commands and arguments must be available using the `-help` flag.
+
+### Process Execution
+
+Jobs will be forked from the service's process and contain exactly one process that it manages. A job must have one of
+the statuses defined in `Statuses` of the [proto spec](../api/proto/job/job.proto) where it will start in an initial
+state and finish in one of three terminal states:
+
+```
+            ┌──► stopped  
+            │             
+running ────┼──► failed   
+            │             
+            └──► succeeded
+```
+
+Resource control will be implemented using cgroups. A parent directory `job-worker` will be created
+under `/sys/fs/cgroup` that will contain a directory for every job created. For example, the path for `job-a` would look
+like `/sys/fs/cgroup/job-worker/job-a`. Under this path, the following files would be updated to manage the resources
+for the job:
+
+* `cgroup.procs` contains the PID of the job's process being managed
+* `cpu.max` limits CPU usage
+* `memory.max` limits memory usage
+* `io.max` limits disk IO
+
+Note that for this implementation, job metadata will remain in memory until the application exits. A real implementation
+of this service would store job metadata in some kind of persistent storage like a database.
+
+### Proto Specification
+
+The service will implement the following protobufs:
+
+* [job](../api/proto/job/job.proto)
+* [worker](../api/proto/worker/worker.proto)
+
+### API
+
+The service's API will contain the resources listed below and be made available using gRPC.
+
+#### worker
+
+The worker is responsible for managing jobs and enforcing authorization for them. The worker may contain many jobs.
+
+* `Query` gets details of jobs owned by the specified user
+
+Details of the worker's functions are listed in [worker.proto](../api/proto/worker/worker.proto).
+
+#### job
+
+A job is responsible for managing the lifecycle of the process given to it. A job only has one worker.
+
+* `Start` start a new job and begin execution of the specified command immediately
+* `Stop` stop execution of the specified job immediately
+* `Query` query details about specified job; this function can run on a job of any status
+* `Output` get the full output (stdout and stderr) of any existing job
```

Comment:
Some more detail here would be nice - how will concurrent output be handled? How will you stream the output from process start? What happens when the job is no longer running?

Reply:
Sounds good. I updated this with a bit more detail, I can dive deeper if it's useful.